From 16eef7644115ba2cd2ccd0438cc26c6eb37efd74 Mon Sep 17 00:00:00 2001
From: Louis Chang <louischang@google.com>
Date: Wed, 25 Dec 2024 11:49:40 +0000
Subject: [PATCH] Avoid app pinning requests if the Task is already locked

Bug: 378514614
Test: LockTaskControllerTest
Flag: EXEMPT bugfix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:b4945cf98c97121821c178713a613a6a6e830c3a)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:a3240bf2ba1587b808b4314dd2e32c624b57c8da)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:326c70b0fd73c3769a252b83e1b3c18069f28b87)
Merged-In: Ic5d2125d39e5e98301d79c8f5aa8986a44e06794
Change-Id: Ic5d2125d39e5e98301d79c8f5aa8986a44e06794
---
 .../core/java/com/android/server/wm/LockTaskController.java  | 4 ++++
 .../src/com/android/server/wm/LockTaskControllerTest.java    | 5 +++++
 2 files changed, 9 insertions(+)

diff --git a/services/core/java/com/android/server/wm/LockTaskController.java b/services/core/java/com/android/server/wm/LockTaskController.java
index 7a055d2948ad..0ea1c3b0c993 100644
--- a/services/core/java/com/android/server/wm/LockTaskController.java
+++ b/services/core/java/com/android/server/wm/LockTaskController.java
@@ -645,6 +645,10 @@ public class LockTaskController {
         if (!isSystemCaller) {
             task.mLockTaskUid = callingUid;
             if (task.mLockTaskAuth == LOCK_TASK_AUTH_PINNABLE) {
+                if (mLockTaskModeTasks.contains(task)) {
+                    Slog.i(TAG_LOCKTASK, "Already locked.");
+                    return;
+                }
                 // startLockTask() called by app, but app is not part of lock task allowlist. Show
                 // app pinning request. We will come back here with isSystemCaller true.
                 ProtoLog.w(WM_DEBUG_LOCKTASK, "Mode default, asking user");
diff --git a/services/tests/wmtests/src/com/android/server/wm/LockTaskControllerTest.java b/services/tests/wmtests/src/com/android/server/wm/LockTaskControllerTest.java
index 1d14dc31fa26..587a2363fca2 100644
--- a/services/tests/wmtests/src/com/android/server/wm/LockTaskControllerTest.java
+++ b/services/tests/wmtests/src/com/android/server/wm/LockTaskControllerTest.java
@@ -239,6 +239,11 @@ public class LockTaskControllerTest {
         verifyLockTaskStarted(STATUS_BAR_MASK_PINNED, DISABLE2_NONE);
         // THEN screen pinning toast should be shown
         verify(mStatusBarService).showPinningEnterExitToast(eq(true /* entering */));
+
+        // WHEN the app calls startLockTaskMode while the Task is already locked
+        mLockTaskController.startLockTaskMode(tr, false, TEST_UID);
+        // THEN a pinning request should NOT be shown
+        verify(mStatusBarManagerInternal, never()).showScreenPinningRequest(anyInt());
     }
 
     @Test
-- 
2.49.0.395.g12beb8f557-goog

